File: binaryflip.py
import random

def flip():
    return random.random() > 0.5

def random_number(n):
    if n == 1:
        return 0
    binary_len = len('{0:0b}'.format(n)) - 1
    
    random_n = ''
    for i in range(binary_len):
        random_n += '1' if flip() else '0'
    
    return int(random_n, 2)


print(random_number(500))
print(random_number(1))
print(random_number(1000))
print(random_number(3))
print(random_number(4))

File: binaryflip.js
function flip() {
    return Math.random() >= 0.5;
}

function randomNumber(n) {
    try {
        if (n > 0 && n < 1000000) {
            if (n === 1) return 0;
            const binaryNumLength = parseInt(n).toString(2).length - 1;
            let randomBinaryNumber = ''
            for (let i = 0; i < binaryNumLength; i++) {
                randomBinaryNumber += flip() ? "1" : "0";
            }
            return parseInt(randomBinaryNumber, 2).toString(10);
        } else {
            throw Error;
        }
    } catch (error) {
        console.error('Invalid input');
    }
}

console.log(randomNumber(500));
console.log(randomNumber(1));
console.log(randomNumber(1000001));
console.log(randomNumber(2));
console.log(randomNumber(500));